src/modules/navigator/mission_block.cpp

```diff
@@ -376,14 +346,62 @@ MissionBlock::is_mission_item_reached()
 	/* Once the waypoint and yaw setpoint have been reached we can start the loiter time countdown */
 	if (_waypoint_position_reached && _waypoint_yaw_reached) {
 
-		if (_time_first_inside_orbit == 0) {
-			_time_first_inside_orbit = now;
-		}
+		bool time_inside_reached = false;
 
 		/* check if the MAV was long enough inside the waypoint orbit */
 		if ((get_time_inside(_mission_item) < FLT_EPSILON) ||
-		    (now - _time_first_inside_orbit >= (hrt_abstime)(get_time_inside(_mission_item) * 1e6f))) {
+		    (now - _time_wp_reached >= (hrt_abstime)(get_time_inside(_mission_item) * 1e6f))) {
+			time_inside_reached = true;
+		}
+
+		// check if heading for exit is reached (only applies for fixed-wing flight)
+		bool exit_heading_reached = false;
+
+		if (time_inside_reached) {
+
+			struct position_setpoint_s *curr_sp_new = &_navigator->get_position_setpoint_triplet()->current;
+			const position_setpoint_s &next_sp = _navigator->get_position_setpoint_triplet()->next;
+
+			/* enforce exit heading if in FW, the next wp is valid, the vehicle is currently loitering and either having force_heading set,
+			   or if loitering to achieve altitdue at a NAV_CMD_WAYPOINT */
+			const bool enforce_exit_heading = _navigator->get_vstatus()->vehicle_type != vehicle_status_s::VEHICLE_TYPE_ROTARY_WING
+							  &&
+							  next_sp.valid &&
+							  curr_sp_new->type == position_setpoint_s::SETPOINT_TYPE_LOITER &&
+							  (_mission_item.force_heading || _mission_item.nav_cmd == NAV_CMD_WAYPOINT);
+
+			if (enforce_exit_heading) {
 
+
+				const float dist_current_next = get_distance_to_next_waypoint(curr_sp_new->lat, curr_sp_new->lon, next_sp.lat,
+								next_sp.lon);
+
+				float yaw_err = 0.0f;
+
+				if (dist_current_next >  1.2f * _navigator->get_loiter_radius()) {
+					// set required yaw from bearing to the next mission item
+					_mission_item.yaw = get_bearing_to_next_waypoint(_navigator->get_global_position()->lat,
+							    _navigator->get_global_position()->lon,
+							    next_sp.lat, next_sp.lon);
+					const float cog = atan2f(_navigator->get_local_position()->vy, _navigator->get_local_position()->vx);
+					yaw_err = wrap_pi(_mission_item.yaw - cog);
+
+
+
+				}
+
+
+				if (fabsf(yaw_err) < 0.1f) { //accept heading for exit if below 0.1 rad error (5.7deg)
```

Comment:
Should this be the parameter (MIS_YAW_ERROR)?

Reply:
True, think that'd be more consistent.

Reply:
addressed in a7a5303bb6a93850c2c373575f5587e15c547dbb